File: app/src/main/java/com/github/sewerina/meter_readings/ui/readings/MainActivity.java
package com.github.sewerina.meter_readings.ui.readings;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.work.PeriodicWorkRequest;
import androidx.work.WorkManager;

import com.github.sewerina.meter_readings.FormattedDate;
import com.github.sewerina.meter_readings.R;
import com.github.sewerina.meter_readings.database.HomeEntity;
import com.github.sewerina.meter_readings.database.ReadingEntity;
import com.github.sewerina.meter_readings.notification.NotificationWorker;
import com.github.sewerina.meter_readings.notification.RemindNotification;
import com.github.sewerina.meter_readings.ui.chart.ChartActivity;
import com.github.sewerina.meter_readings.ui.homes.HomesActivity;
import com.github.sewerina.meter_readings.ui.settings.SettingsActivity;
import com.google.android.material.floatingactionbutton.FloatingActionButton;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = "MainActivity";
    public static final String CURRENT_HOME_ENTITY = "currentHomeEntity";

    private Spinner mSpinner;
    private RecyclerView mRecyclerView;
    private FloatingActionButton mAddReadingFab;

    private ArrayAdapter<String> mSpinnerAdapter;
    private ReadingAdapter mReadingAdapter;

    private MainViewModel mViewModel;
    private AdapterView.OnItemSelectedListener mOnItemSelectedListener = new AdapterView.OnItemSelectedListener() {
        @Override
        public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
            mViewModel.changeCurrentHome(position);
        }

        @Override
        public void onNothingSelected(AdapterView<?> parent) {

        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mViewModel = ViewModelProviders.of(this).get(MainViewModel.class);

        if (savedInstanceState != null) {
            HomeEntity homeEntity = (HomeEntity) savedInstanceState.getSerializable(CURRENT_HOME_ENTITY);
            mViewModel.restore(homeEntity);
        }
        mViewModel.getReadings().observe(this, new Observer<List<ReadingEntity>>() {
            @Override
            public void onChanged(List<ReadingEntity> readingEntities) {
                mReadingAdapter.update(readingEntities);
            }
        });

        mViewModel.getState().observe(this, new Observer<MainViewModel.State>() {
            @Override
            public void onChanged(MainViewModel.State state) {
                mSpinnerAdapter.clear();
                List<HomeEntity> homeEntityList = state.homeEntityList;
                for (HomeEntity homeEntity : homeEntityList) {
                    mSpinnerAdapter.add(homeEntity.address);
                }
                mSpinnerAdapter.notifyDataSetChanged();
                mSpinner.setSelection(state.currentHomePosition);
            }
        });

        mSpinner = findViewById(R.id.spinner_home);
        mSpinnerAdapter = new ArrayAdapter<>(MainActivity.this,
                android.R.layout.simple_spinner_item, new ArrayList<String>());
        mSpinnerAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        mSpinner.setAdapter(mSpinnerAdapter);
        mSpinner.setOnItemSelectedListener(mOnItemSelectedListener);

        mReadingAdapter = new ReadingAdapter();
        mRecyclerView = findViewById(R.id.recyclerReadings);
//        mRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        mRecyclerView.setLayoutManager(new GridLayoutManager(this, 2));
//        mRecyclerView.setLayoutManager(new GridLayoutManager(this, 2, RecyclerView.VERTICAL, false));
        mRecyclerView.setAdapter(mReadingAdapter);

        mAddReadingFab = findViewById(R.id.fab_addReading);
        mAddReadingFab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                NewReadingDialog.showDialog(getSupportFragmentManager());
            }
        });

//        mViewModel.firstLoad();
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        if (mViewModel.getState().getValue() != null) {
            outState.putSerializable(CURRENT_HOME_ENTITY, mViewModel.getState().getValue().currentHomeEntity);
        }
        super.onSaveInstanceState(outState);
    }

    @Override
    protected void onResume() {
        super.onResume();
        mViewModel.load();
//        new RemindNotification(this).appearNotification();
        PeriodicWorkRequest notificationWorkRequest = new PeriodicWorkRequest
                .Builder(NotificationWorker.class, 1, TimeUnit.HOURS)
                .build();
        WorkManager.getInstance(this).enqueue(notificationWorkRequest);
    }

    @Override
    protected void onPause() {
        super.onPause();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        if (id == R.id.action_homes) {
            startActivity(new Intent(this, HomesActivity.class));
            return true;
        }

        if (id == R.id.action_settings) {
            startActivity(new Intent(this, SettingsActivity.class));
            return true;
        }

        if (id == R.id.action_chart) {
            if (mViewModel.getState().getValue() != null) {
                startActivity(ChartActivity.newIntent(this, mViewModel.getState().getValue().currentHomeEntity));
                return true;
            }
        }

        return super.onOptionsItemSelected(item);
    }

    private class ReadingHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        private TextView mDateTv;
        private TextView mColdWaterTv;
        private TextView mHotWaterTv;
        private TextView mDrainWaterTv;
        private TextView mElectricityTv;
        private TextView mGasTv;

        private ReadingEntity mReadingEntity;

        public ReadingHolder(@NonNull View itemView) {
            super(itemView);
            itemView.setOnClickListener(this);

            new ReadingPreferences(itemView.getContext()).setLayoutVisibility(itemView);

            mDateTv = itemView.findViewById(R.id.tv_date);
            mColdWaterTv = itemView.findViewById(R.id.tv_coldWater);
            mHotWaterTv = itemView.findViewById(R.id.tv_hotWater);
            mDrainWaterTv = itemView.findViewById(R.id.tv_drainWater);
            mElectricityTv = itemView.findViewById(R.id.tv_electricity);
            mGasTv = itemView.findViewById(R.id.tv_gas);
        }

        void bind(ReadingEntity entity) {
            mReadingEntity = entity;

            mDateTv.setText(new FormattedDate(entity.date).text());
            mColdWaterTv.setText(String.valueOf(entity.coldWater));
            mHotWaterTv.setText(String.valueOf(entity.hotWater));
            mDrainWaterTv.setText(String.valueOf(entity.drainWater));
            mElectricityTv.setText(String.valueOf(entity.electricity));
            mGasTv.setText(String.valueOf(entity.gas));
        }

        @Override
        public void onClick(View v) {
            Log.d(TAG, "onClick: ");
            BottomSheetReadingDialog.showDialog(getSupportFragmentManager(), mReadingEntity);
        }
    }

    private class ReadingAdapter extends RecyclerView.Adapter<ReadingHolder> {
        private final List<ReadingEntity> mReadings = new ArrayList<>();

        @NonNull
        @Override
        public ReadingHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.reading_list_item, parent, false);
            return new ReadingHolder(itemView);
        }

        @Override
        public void onBindViewHolder(@NonNull ReadingHolder holder, int position) {
            holder.bind(mReadings.get(position));
        }

        @Override
        public int getItemCount() {
            return mReadings.size();
        }

        void update(List<ReadingEntity> entities) {
            mReadings.clear();
            mReadings.addAll(entities);
            notifyDataSetChanged();
        }

        void add(ReadingEntity entity) {
            mReadings.add(entity);
            notifyItemInserted(mReadings.size() - 1);
        }
    }


}

File: app/src/main/java/com/github/sewerina/meter_readings/ui/readings_main/MainActivity.kt
package com.github.sewerina.meter_readings.ui.readings_main

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.*
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Spinner
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.github.sewerina.meter_readings.FormattedDate
import com.github.sewerina.meter_readings.R
import com.github.sewerina.meter_readings.ReadingApp
import com.github.sewerina.meter_readings.database.HomeEntity
import com.github.sewerina.meter_readings.database.ReadingEntity
import com.github.sewerina.meter_readings.ui.backup_copying.BackupCopyingActivity
import com.github.sewerina.meter_readings.ui.chart.ChartActivity
import com.github.sewerina.meter_readings.ui.homes.HomesActivity
import com.github.sewerina.meter_readings.ui.report.ReportActivity
import com.github.sewerina.meter_readings.ui.settings.SettingsActivity
import com.google.android.material.floatingactionbutton.FloatingActionButton
import javax.inject.Inject

class MainActivity : AppCompatActivity() {
    private lateinit var mSpinner: Spinner
    private lateinit var mRecyclerView: RecyclerView
    private lateinit var mAddReadingFab: FloatingActionButton
    private lateinit var mSpinnerAdapter: ArrayAdapter<String>
    private lateinit var mReadingAdapter: ReadingAdapter

    @Inject
    lateinit var mViewModel: MainViewModel

    private val mOnItemSelectedListener: AdapterView.OnItemSelectedListener =
        object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(
                parent: AdapterView<*>?,
                view: View,
                position: Int,
                id: Long
            ) {
                mViewModel.changeCurrentHome(position)
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        ReadingApp.sMainComponent?.inject(this)

        mSpinner = findViewById(R.id.spinner_home)
        mSpinnerAdapter =
            ArrayAdapter(this@MainActivity, android.R.layout.simple_spinner_item, ArrayList())
        mSpinnerAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        mSpinner.apply {
            adapter = mSpinnerAdapter
            onItemSelectedListener = mOnItemSelectedListener
        }

        mReadingAdapter = ReadingAdapter()
        mRecyclerView = findViewById<RecyclerView>(R.id.recyclerReadings).apply {
            layoutManager = GridLayoutManager(this@MainActivity, 2)
            adapter = mReadingAdapter
        }

        mAddReadingFab = findViewById(R.id.fab_addReading)
        mAddReadingFab.setOnClickListener {
            NewReadingDialog.showDialog(supportFragmentManager)
        }

        if (savedInstanceState != null) {
            val homeEntity = savedInstanceState.getSerializable(CURRENT_HOME_ENTITY) as HomeEntity
            mViewModel.restore(homeEntity)
        }

        mViewModel.readings.observe(this) { readingEntities ->
            mReadingAdapter.update(readingEntities)
        }

        mViewModel.state.observe(this) { state ->
            mSpinnerAdapter.clear()
            val homeEntityList = state!!.homeEntityList
            for (homeEntity in homeEntityList) {
                mSpinnerAdapter.add(homeEntity.address)
            }
            mSpinnerAdapter.notifyDataSetChanged()
            mSpinner.setSelection(state.currentHomePosition)
        }

//        mViewModel.firstLoad();
    }

    override fun onSaveInstanceState(outState: Bundle) {
        if (mViewModel.state.value != null) {
            outState.putSerializable(
                CURRENT_HOME_ENTITY,
                mViewModel.state.value!!.currentHomeEntity
            )
        }
        super.onSaveInstanceState(outState)
    }

    override fun onResume() {
        super.onResume()
        mViewModel.load()
    }

    override fun onPause() {
        super.onPause()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.action_homes) {
            startActivity(Intent(this, HomesActivity::class.java))
            return true
        }
        if (id == R.id.action_chart) {
            startActivity(
                mViewModel.state.value!!.currentHomeEntity?.let {
                    ChartActivity.newIntent(this, it)
                }
            )
            return true
        }
        if (id == R.id.action_settings) {
            startActivity(Intent(this, SettingsActivity::class.java))
            return true
        }
        if (id == R.id.action_backupCopying) {
            startActivity(Intent(this, BackupCopyingActivity::class.java))
            return true
        }
        if (id == R.id.action_reports) {
            startActivity(
                mViewModel.state.value!!.currentHomeEntity?.let {
                    ReportActivity.newIntent(this, it)
                }
            )
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private inner class ReadingHolder(itemView: View) : RecyclerView.ViewHolder(itemView),
        View.OnClickListener {
        private val mDateTv: TextView
        private val mColdWaterTv: TextView
        private val mHotWaterTv: TextView
        private val mDrainWaterTv: TextView
        private val mElectricityTv: TextView
        private val mGasTv: TextView

        private lateinit var mReadingEntity: ReadingEntity
        fun bind(entity: ReadingEntity) {
            mReadingEntity = entity
            ReadingPreferences().setLayoutVisibility(itemView)
            mDateTv.text = FormattedDate(entity.date).text()
            mColdWaterTv.text = entity.coldWater.toString()
            mHotWaterTv.text = entity.hotWater.toString()
            mDrainWaterTv.text = entity.drainWater.toString()
            mElectricityTv.text = entity.electricity.toString()
            mGasTv.text = entity.gas.toString()
        }

        override fun onClick(v: View) {
            Log.d(TAG, "onClick: ")
            BottomSheetReadingDialog.showDialog(supportFragmentManager, mReadingEntity)
        }

        init {
            itemView.setOnClickListener(this)
            mDateTv = itemView.findViewById(R.id.tv_date)
            mColdWaterTv = itemView.findViewById(R.id.tv_coldWater)
            mHotWaterTv = itemView.findViewById(R.id.tv_hotWater)
            mDrainWaterTv = itemView.findViewById(R.id.tv_drainWater)
            mElectricityTv = itemView.findViewById(R.id.tv_electricity)
            mGasTv = itemView.findViewById(R.id.tv_gas)
        }
    }

    private inner class ReadingAdapter : RecyclerView.Adapter<ReadingHolder>() {
        private val mReadings: MutableList<ReadingEntity> = ArrayList()
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ReadingHolder {
            val itemView = LayoutInflater.from(parent.context)
                .inflate(R.layout.reading_list_item, parent, false)
            return ReadingHolder(itemView)
        }

        override fun onBindViewHolder(holder: ReadingHolder, position: Int) {
            holder.bind(mReadings[position])
        }

        override fun getItemCount(): Int {
            return mReadings.size
        }

        fun update(entities: List<ReadingEntity>?) {
            mReadings.clear()
            mReadings.addAll(entities!!)
            notifyDataSetChanged()
        }

        fun add(entity: ReadingEntity) {
            mReadings.add(entity)
            notifyItemInserted(mReadings.size - 1)
        }
    }

    companion object {
        const val CURRENT_HOME_ENTITY = "currentHomeEntity"
        private const val TAG = "MainActivity"
    }
}